Fix #51: use stdin for ruff

densnow opened the pull request:
Ruff  uses the `-` option in combination with `subprocess.run`'s `input=str` to read directly from `str` through stdin.

Tested on the example essays, including the coding guide, and TMA's 1, 2 and 3, with and without `notebook_cell.py` files in the same directory.

Additionally made minor modifications to the docstring of `load_ipython_extension`

densnow commented:
I've added some infrastructure to automate testing for `%ruff` and `%allowed` and added a few tests. 

The idea is to yield an Ipython `InteractiveShell` fixture, with the algoesup extension loaded, for each test. The cleanup/teardown then resets the state of the shell ready for the next test.

During the tests we use the shells `run_cell` method in combination with `capture_output` to run and capture output  ready for testing against expected.

If you can think of any important tests I've forgotten, please let me know and I'll add them.

mwermelinger commented:
The plot thickens: if I run the coding.ipynb notebook, which is the only one with ruff errors, then everything's fine, i.e. run ruff reports the errors instead of saying it can't parse the .toml file.

densnow commented:
So it's the `pyproject.toml` local to `src/docs/essays` that's causing the issue? 

What version of `ruff` are you using? I can remember a while back getting deprecation warnings about the naming conventions in the pyproject file. I can't seem to find it in the revision history though. 

All that is to say that I think this has not always been the case, but is to do with newer versions of `ruff`.

densnow commented:
I think this is probably explained in the [config file discovery section of the ruff docs](https://docs.astral.sh/ruff/configuration/#:~:text=see%20Settings.-,Config%20file%20discovery,you%20to%20inherit%20the%20settings%20from%20another%20config%20file%2C%20like%20so%3A,-pyproject.toml). 

After a quick skim they say that ruff searches for the "closest config file in the file hierarchy" and if none is found uses a default configuration.

mwermelinger commented:
Thanks. Since we might have to do changes anyhow due to JupyterLab, my preference for the moment is just to remove the `output-format` line from the .toml file and keep everything else as is.

mwermelinger commented:
I don't plan anything further on this one, so I'll merge unless you still want to do something.

densnow commented:
I don't necessarily want to add any more to this, so its fine to go ahead and merge. 

I think in the long run adding a few more tests for `%allowed` (like testing with some other options) would provide better coverage, but I lost the motivation to add more currently.

densnow commented:
Changed my mind, and made a couple of last minute changes to tidy things up:
- Used list comprehension instead of for loop for `no_e501_on_transformed`, as they are better optimised for this task
- changed a couple of comments etc

Definitely finished now though, so merge if you are happy with everything :+1:

mwermelinger reviewed (approved):
Sorry for the huge delay. Have run the tests, an essay and coding.ipynb and all seems well. Many thanks for this.
